kroxylicious-filters/kroxylicious-record-encryption/src/main/java/io/kroxylicious/filter/encryption/encrypt/EncryptionDekCache.java

```diff
@@ -120,12 +126,22 @@ private void afterCacheEviction(@Nullable CacheKey<K> cacheKey,
     }
 
     /**
-     * Discard any cached DEK for the KEK in the given {@code encryptionScheme}.
+     * Discard the cached DEK for the KEK in the given {@code encryptionScheme}. This uses a
+     * CAS style so that multiple threads can redundantly invalidate the same cache key.
      * This method may block if a DEK for the given {@code encryptionScheme} is in the process
      * of being loaded.
      * @param encryptionScheme The KEK for the DEK to discard.
+     * @param dek The DEK to discard.
      */
-    public void invalidate(@NonNull EncryptionScheme<K> encryptionScheme) {
-        dekCache.synchronous().invalidate(cacheKey(encryptionScheme));
+    public void invalidate(@NonNull EncryptionScheme<K> encryptionScheme, @NonNull Dek<E> dek) {
+        dekCache.asMap().computeIfPresent(cacheKey(encryptionScheme), (kCacheKey, dekCompletableFuture) -> {
+            if (dekCompletableFuture.isDone() && !dekCompletableFuture.isCompletedExceptionally() && dekCompletableFuture.join() == dek) {
+                invalidationCount.incrementAndGet();
+                return null;
+            }
+            else {
```

Comment:
should we have some logging or metrics around the main clause?

Hang on why do we return null if `dekCompletableFuture.join() == dek` ?

Reply:
The aim is to remove the mapping from the cache if the current value in the cache is this dek. `ConcurrentMap#computeIfPresent` is docced as `If the remapping function returns null, the mapping is removed.`

So multiple threads that got a hold of the same DEK instance can all call this code and it should only result in a single removal of the DEK.

If a thread observes a non-done or exceptionally completed future it doesn't remove it as it is in flight or failed. If a thread observes a different DEK from the one it's trying to expire, then it doesn't remove it, it's probably the next DEK.

Reply:
I think its probably worth adding a version of 
>ConcurrentMap#computeIfPresent is docced as If the remapping function returns null, the mapping is removed.

as a comment around the code as I didn't recognise thats what was happening when reading it. Otherwise it sounds perfectly reasonable to me.